### ISSUE TITLE
Ampersand in TrackInfo data causes crash

### ISSUE THREAD

**Issue by grahamdw** [bazwilliams/openhomedevice#16]
If track metadata contains an ampersand, then the call _metadataXml = etree.fromstring(self.metadata)_ fails with an error of **xml.etree.ElementTree.ParseError: not well-formed (invalid token)**. I suspect this is due to the entire value of **metadata** getting unescaped when the response XML is parsed. The markup contained in **metadata** needs to be unescaped, but the data should not be (this looks messy to achieve!)

Similar errors perhaps happen if other escaped XML characters are in the track metadata (I would expect  **<** at least to cause problems).

**Comment by bazwilliams** [comment]
Can you provide an example please? The test data in the repo has an ampersand in the title and that seems to work as does a track played from Spotify. 

Could you run device spy and call the info service “Track” method with the offending track so I can investigate please?

You could also put a print statement before the parse in this library to get the raw response if you can?

**Comment by grahamdw** [comment]
Thanks for the quick response. I've now tracked down my problem - I had some software that was preparing playlists, and it didn't double escape ampersands when preparing the metadata value in the XML it generated. So the single escaped ampersands then caused a parsing error on the second parsing. Sorry to have troubled you...

**Comment by bazwilliams** [comment]
Cool. Looks like there’s a bug anyway, it should probably handle the metadata string more gracefully if not escaped. Feel free to send a sample and I’ll add some error handling.

**Comment by grahamdw** [comment]
Example xml with single escaped ampersand in **genre** value.

[example.xml.txt](https://github.com/bazwilliams/openhomedevice/files/3946573/example.xml.txt)

One other (unrelated) thing crossed my mind, about Soap.py. I'm using openhomedevice from Home Assistant, and I've occasionally had problems when the device it's talking to has gone offline. Home Assistant has tried to access it, presumably timed out the thread, and then restarted it (repeats until the device is reset). Would it be possible to have a timeout for the request post itself, after say 20 seconds or so?

**Comment by bazwilliams** [comment]
Thanks for the data. I’ll see about improving the robustness. 

Awesome about HA use, I’m a fan and use it too! It’s an interesting case you have, I’ll see what can be done. Because the HA integration uses local polling I think HA will just ask repeatedly, I think there may be rules about timeouts too so will ask on their forums. Either way again the lib should handle device offline more elegantly. Better yet, we change it to local push notifications on a upnp subscription.

**Comment by grahamdw** [comment]
Thanks - that'd be good. I've got the playlist software changed now to
double escape ampersands, so hopefully the need will be rather less!

I really like HA too. I've been using it for a couple of years now, and
have got lots of devices connected and automated. I'm really grateful for
the openhomedevice module - I have 3 media players in different rooms. HA
with your code means I can easily do things like when the kitchen player is
started it will automatically play what the living room player is playing,
or if the living room player is not running, it will use its own playlist.
The only bugbear has been when HA can't detect correctly what state a
player is in, and thanks to your response I've got rid of a whole lot of
those cases.

I had in mind something simple for the timeout - just adding a timeout
parameter to the *requests.post *call in Soap.py, so it would become (eg)
*res = requests.post(location, data=bodyString, headers=headers,
timeout=(2,20) )*
(disclaimer - Python is not my first language - I'm just picking it up as I
need it)
There'd be knock on effects to be handled when Timeout is raised, but if
I've understood requests.post correctly, it waits indefinitely if a timeout
is not specified?
